### PR TITLE
Cleanup NumberFormatInfo tests

### PR DESCRIPTION
**Pull request by hughbe**
Contributes to #2336

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
Test Innerloop OSX Debug Build and Test please

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyDecimalDigits.cs [hunk 1]
```diff
@@ -11,80 +12,38 @@ namespace System.Globalization.Tests
...
         {
-            NumberFormatInfo nfi = new NumberFormatInfo();
-            NumberFormatInfo nfiReadOnly = NumberFormatInfo.ReadOnly(nfi);
-            Assert.Throws<InvalidOperationException>(() =>
-            {
-                nfiReadOnly.CurrencyDecimalDigits = 1;
-            });
+            yield return new object[] { NumberFormatInfo.InvariantInfo, 2, 2 };
+            yield return new object[] { new CultureInfo("en-US").NumberFormat, 2, 2 };
+            yield return new object[] { new CultureInfo("ko").NumberFormat, 0, 2 }; // Todo: determine why some values are different
```
Comment (stephentoub) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyDecimalDigits.cs]: Is there an issue associated with this TODO?  If no, could you open one?  Either way, we should include the issue # here.
cc: @eerhardt

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyDecimalSeparator.cs [hunk 2]
```diff
@@ -2,68 +2,34 @@
...
         {
-            string testStr = "";
-            NumberFormatInfo nfi = new NumberFormatInfo();
-            NumberFormatInfo nfiReadOnly = NumberFormatInfo.ReadOnly(nfi);
-            Assert.Throws<ArgumentException>(() =>
-            {
-                nfi.CurrencyDecimalSeparator = testStr;
-            });
+            Assert.Throws<ArgumentNullException>(() => new NumberFormatInfo().CurrencyDecimalSeparator = null);
+            Assert.Throws<ArgumentException>(() => new NumberFormatInfo().CurrencyDecimalSeparator = "");
```
Comment (stephentoub) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyDecimalSeparator.cs]: Can you include the parameter name in these and other arg exceptions?

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyDecimalDigits.cs (hunk 1 above)
Comment (eerhardt) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyDecimalDigits.cs]: Looks like @steveharter added these TODOs. In general, ICU data and Windows data does different in many cases, so having the test account for that makes sense.

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyNegativePattern.cs [hunk 3]
```diff
@@ -2,95 +2,51 @@
...
-            });
+            yield return new object[] { NumberFormatInfo.InvariantInfo, new int[] { 0 } };
+            yield return new object[] { new CultureInfo("bg-BG").NumberFormat, new int[] { 8 } };
+            yield return new object[] { new CultureInfo("en-US").NumberFormat, NumberFormatInfoData.GetCurrencyNegativePatterns("en-US") };
+            yield return new object[] { new CultureInfo("en-CA").NumberFormat, NumberFormatInfoData.GetCurrencyNegativePatterns("en-CA") };
+            yield return new object[] { new CultureInfo("fa-IR").NumberFormat, NumberFormatInfoData.GetCurrencyNegativePatterns("fa-IR") };
+            yield return new object[] { new CultureInfo("fr-CD").NumberFormat, NumberFormatInfoData.GetCurrencyNegativePatterns("fr-CD") };
+            yield return new object[] { new CultureInfo("as").NumberFormat, NumberFormatInfoData.GetCurrencyNegativePatterns("as") };
+            yield return new object[] { new CultureInfo("es-BO").NumberFormat, NumberFormatInfoData.GetCurrencyNegativePatterns("es-BO") };
+            yield return new object[] { new CultureInfo("fr-CA").NumberFormat, NumberFormatInfoData.GetCurrencyNegativePatterns("fr-CA") };
```
Comment (stephentoub) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyNegativePattern.cs]: Why is this transformation better than the original?  Seems like lots of duplicated code that the original didn't have.

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoData.cs [hunk 4]
```diff
@@ -2,30 +2,21 @@
...
-using Xunit;
-
 namespace System.Globalization.Tests
 {
     internal static class NumberFormatInfoData
     {
-        private static readonly bool s_isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
-        private static readonly int s_WindowsVersion = PlatformDetection.WindowsVersion;
-        private static readonly bool s_isOSX = RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
-        private static readonly Version s_OSXKernelVersion = GetOSXKernelVersion();
```
Comment (stephentoub) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoData.cs]: Why did you remove these and replace all call sites with the expanded code?

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyDecimalDigits.cs (hunk 1 above)
Comment (steveharter) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoCurrencyDecimalDigits.cs]: The TODO can be removed. ICU had some values that didn't seem right at the time but after spending more time looking at other inconsistencies between windows and ICU the results are expected and don't require further research.

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoPercentNegativePattern.cs [hunk 5]
```diff
@@ -11,80 +10,40 @@ namespace System.Globalization.Tests
...
+        public static IEnumerable<object[]> PercentNegativePattern_TestData()
         {
-            NumberFormatInfo nfi = new NumberFormatInfo();
-            for (int i = 0; i <= 11; i++)
-            {
-                nfi.PercentNegativePattern = i;
-                Assert.Equal(i, nfi.PercentNegativePattern);
-            }
+            yield return new object[] { NumberFormatInfo.InvariantInfo, 0, 0 };
+            yield return new object[] { new CultureInfo("en-US").NumberFormat, 0, 1 }; // Todo: determine why some values are different
```
Comment (steveharter) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoPercentNegativePattern.cs]: Please remove this and similar Todos

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoNumberNegativePattern.cs [hunk 6]
```diff
@@ -2,74 +2,44 @@
...
+            yield return new object[] { NumberFormatInfo.InvariantInfo, 1 };
+            yield return new object[] { new CultureInfo("en-US").NumberFormat, 1 };
         }
 
-        // TestSetValue: Verify set value of property NumberNegativePattern
-        [Fact]
-        public void TestSetValue()
+        [Theory]
+        [MemberData(nameof(NumberNegativePattern_TestData))]
+        public void NumberNegativePattern(NumberFormatInfo format, int expected)
```
Comment (stephentoub) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoNumberNegativePattern.cs]: Shouldn't this be named "NumberNegativePattern_Get" or something like that to mirror the set version below?

File: src/System.Globalization/tests/PlatformDetection.cs [hunk 7]
```diff
@@ -13,18 +15,7 @@ public static class PlatformDetection
...
-        get
-        {
-            if (!s_WindowsVersion.HasValue)
-            {
-                s_WindowsVersion = GetWindowsVersion();
-            }
-            return s_WindowsVersion.Value;
-        }
-    }
+    public static int WindowsVersion => GetWindowsVersion();
```
Comment (stephentoub) [inline src/System.Globalization/tests/PlatformDetection.cs]: You removed the caching from this. Why?

File: src/System.Globalization/tests/PlatformDetection.cs [hunk 8]
```diff
@@ -51,6 +42,8 @@ public static bool IsUbuntu1510
         }
     }
 
+    public static Version OSXKernelVersion => GetOSXKernelVersion();
```
Comment (stephentoub) [inline src/System.Globalization/tests/PlatformDetection.cs]: This should be cached.

**Comment by stephentoub** [comment]
Some issues to be addressed, but otherwise LGTM.

cc: @tarekgh

**Comment by tarekgh** [comment]
other than @stephentoub issues and comments, LGTM

I know the test logic is not changed here but in general I am still think we shouldn't hardcore culture data in our tests especially with cultures that can likely change. I know we test the OS version, but this doesn't guarantee the test will always succeed in cases if the OS get patched or in Windows someone install replacement custom culture. I know we have discussed this before and we didn't reach conclusion but I wanted to say that just to consider for future tests

File: src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoData.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoData.cs]: I didn't really think that creating a duplicate variable for something that is already cached(/will be when I undo my changes). There's no real change tbh

**Comment by hughbe** [comment]
@tarekgh I actually think that these tests are duplicated. Here's my reasoning:
There's a test caused CultureInfoAll.cs in the project that (I think) takes every single system culture and makes sure that {get} property is properly filled out.
Let me know what you think

**Comment by tarekgh** [comment]
@hughbe first CultreInfoAll is Windows specific and not running on Linux/OSX. second CultureInfoAll exercise the data only but it would be good to have tests to check the final results used by this data. for example, if the default short date format for en-US is mm/dd/yyyy, it would be good to check the formatted date (like 03/14/2016) and not only check the data "mm/dd/yyyy". so I believe we still need such tests we have here.

**Comment by hughbe** [comment]
Rebased to fix another merge conflict
